### ISSUE TITLE
面试知识点复习-四月版

### ISSUE THREAD

**Issue by Eamymao** [Eamymao/winter-ann-arbor#14]
- [ ] 1 你的研究方向是啥？
- [ ] 2 说说你的研究方向
- [ ] 3 聊了半天项目

- [ ] 最近读了什么paper，有什么收获，介绍一下？

----------------------------------------------------------

- [ ] SVM适用于什么场景，或者有什么限制？
LR和SVM、线性回归的联系与区别
> https://blog.csdn.net/sinat_32043495/article/details/79681177

- [ ] kmeans算法的细节，优缺点？
----------------------------------------------------------
### LR回归相关
- [x]  CTR中为什么经常用LR？
> 为什么LR可以用来做CTR预估？https://www.zhihu.com/question/23652394

1. CTR需要把被点击的样本当成正例，把未点击的样本当成负例，那么样本的ctr实际上就是样本为正例的概率，LR可以输出样本为正例的概率，所以可以用来解决这类问题.
2. 广告点击率预测需要基于大规模的训练样本和特征。一般是十亿级别样本和亿级别的特征，甚至更多。逻辑回归由于模型简单，很适合用来学习大数据的问题, 处理速度也快.
3. 是参数模型, 可解释性强,每个特征有对应的参数,  不同的参数说明了不同的特征对CTR的影响大小.

- [x] LR中参数的意义，如何检验？p-value的意义？
$$\text{Logit}(p)=\log\left(\frac{p}{1-p}\right)=\beta_0+\beta_1x_1+\beta_2x_2+\cdots+\beta_px_p$$
Logit(p)和变量的系数是线性关系，也就是log odds(odds即事件发生的概率与不发生的概率之比)和系数是成线性关系。也就是在逻辑回归当中，x每增加一个单位，logit函数增加beta个单位。
>实际医学案例来说明Logistic回归模型参数的含义: https://zhuanlan.zhihu.com/p/49481076

似然比（Likelihood ratio test）检验
在一个模型里面，含有变量x与不含变量x的对数似然值乘以-2的结果之差，服从卡方分布。这一检验统计量称为似然比(likelihood ratio)，用式子表示为G = -2ln(不含x似然/含有x似然).
计算统计量G在卡方分布中对应的概率值(即P(x>G)),如果P小于p=0.05,即G落在拒绝域中,拒绝零假设,变量x的系数不为0.
> LR相关的检验: https://www.cnblogs.com/biyeymyhjob/archive/2012/07/18/2595410.html

- [x] LR里面，损失函数能不能把交叉熵换成MSE？

不好,因为交叉熵形式的最大似然函数求出来的梯度不含有sigmoid项的梯度且只与x和y相关,但是MSE的梯度表达式中会含有sigmoid项的梯度,sigmoid项的梯度在定义域内小于0.25,会导致收敛速度变慢.

- [x] LR与线性回归的比较？LR的前提假设是什么？

> 逻辑回归的常见面试点总结: https://www.cnblogs.com/modifyrong/p/7739955.html

> 逻辑回归(logistic regression)的本质——极大似然估计: https://blog.csdn.net/zjuPeco/article/details/77165974

LR得到的是0-1之间的概率值,线性回归则是在整个实数域范围内进行预测.
经典线性模型的优化目标函数是最小二乘，而逻辑回归则是似然函数

----------------------------------------------------------
### GBDT,XGB,RF树相关
- [x] GBDT的特征组合原理

> GBDT原理及利用GBDT构造新的特征 https://blog.csdn.net/shine19930820/article/details/71713680

> 作者总结: 对于样本量大的数据，线性模型具有训练速度快的特点，但线性模型学习能力限于线性可分数据，所以就需要特征工程将数据尽可能地从输入空间转换到线性可分的特征空间。GBDT与LR的融合模型，其实使用GBDT来发掘有区分度的特征以及组合特征，来替代人工组合特征。

用已有特征训练GBDT模型，然后利用GBDT模型学习到的树来构造新特征，最后把这些新特征加入原有特征一起训练模型。构造的新特征向量是取值0/1的，向量的每个元素对应于GBDT模型中树的叶子结点。当一个样本点通过某棵树最终落在这棵树的一个叶子结点上，那么在新特征向量中这个叶子结点对应的元素值为1，而这棵树的其他叶子结点对应的元素值为0。新特征向量的长度等于GBDT模型里所有树包含的叶子结点数之和。

- [ ] 决策树相比其他算法有什么优势？
- [ ] 决策树分裂节点的选择
- [x] gbdt 和 xgb

> RF、GBDT、XGBoost常见面试题整理 https://zhuanlan.zhihu.com/p/34679467

- [x] 为什么项目选择随机森林而不选择GBDT？

If you carefully tune parameters, gradient boosting can result in better performance than random forests. However, gradient boosting may not be a good choice if you have a lot of noise, as it can result in overfitting. They also tend to be harder to tune than random forests.

Random forests and gradient boosting each excel in different areas. Random forests perform well for multi-class object detection and bioinformatics, which tends to have a lot of statistical noise. Gradient Boosting performs well when you have unbalanced data such as in real time risk assessment.

都尝试过,但是表现比较相近. 样本量不大,特征也较少,xgboost可能会过拟合
- [x] GBDT/xgboost调参

> 调优指南: https://blog.csdn.net/u010665216/article/details/78532619
min_child_weight means something like "stop trying to split once you reach a certain degree of purity in a node and your model can fit it".

- [x] 方差偏差角度解释bagging，为什么随机森林泛化能力强

> 为什么说bagging是减少variance，而boosting是减少bias?https://www.zhihu.com/question/26760839

- [ ] boosting算法，adaboost是怎么更新的

----------------------------------------------------------
- [ ] 1、自我介绍
- [ ] 2、细扣实习项目（问的比较深）
- [ ] 3、数据清洗、数据平滑主要操作
- [ ] 11、最大似然估计和贝叶斯估计的联系和区别
- [ ] 12、如果我想预测的结果是一个置信区间你打算怎么建模型（这个是在最前面的问题）
----------------------------------------------------------
### 梯度下降
- [ ] 5、详细说一下随机梯度下降和批量梯度下降
- [ ] 6、线性回归多变量求解的过程，为什么这样求解？这样求解为什么是最优解？（我回答的是求导，也就是梯度下降）
- [ ] 7、怎么优化梯度下降过程，主要是速度优化？
- [ ] 8、自适应梯度优化是什么样子？
- [ ] 9、拟牛顿法能说说吗？
- [ ] 10、学习率过大会出现什么问题，怎么解决
- [ ] 11、一阶导和二阶导分别表示什么，能说一下意义吗？/数学角度解释sgd
- [ ] 常用优化器

----------------------------------------------------------
- [x] 1. NLP哪个模型最熟悉？Word2Vec
- [ ] 2. Word2vec的计算优化方法: 分层softmax, 负采样
- [x] 3. Word2vec数学原理
- [ ] 6. ReLu函数所有背景、原理、应用；还有什么激活函数？不同激活函数的对比？
- [ ] 8. TF-IDF背景、原理、应用、优缺点
- [x] 9. bert当时是用预训练模型还是自己重新训练的？用bert来干嘛了？
- [ ] 10. n-gram算法原理？这里用n-gram作用？这个最大长度匹配是干嘛的？
- [x] 11. DL常用模型还知道啥？（补RNN、bert、transformer）
- [ ] 深度学习里面解决梯度消失的办法
----------------------------------------------------------

- [ ] 4. Dropout缺点
ps: 这个回答成Relu的缺点了，不过面试官也没有提出疑问，直接下一个问题了
- [ ] 5. 讲一下NLP中有哪些数据增强方法
- [ ] 6. 讲一下NLP中都有哪些种类的Attention

- [ ] lstm的结构

----------------------------------------------------------
数据预处理怎么做，怎么数据清洗，类别不均衡怎么处理的，模型怎么构建
有一个人你知道他银行卡金额上限和下限，猜银行卡金额，猜对了就把金额给你，想要概率最大（猜的次数最少），金额最大，可以猜无数次，怎么优化，怎么建模，可以先做一些实验辅助。
Word2Vec 怎么将得到的词向量变成句子向量，以及怎么衡量 Word2Vec得到词向量的好坏，这个我总结到了我的博客 https://blog.csdn.net/Matrix_cc/article/details/105138478

----------------------------------------------------------
### **算法题**
- [x] 6 做了个题目，
https://zhuanlan.zhihu.com/p/112002408
https://leetcode-cn.com/problems/longest-common-subsequence/
- [ ] 7 什么情况下用动态规划，如果换成递归会怎么样
- [x] 8. 算法题一，买入股票问题，只能买卖一次求最大收益
https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-iii/
- [x] 9.  算法题二，01背包
https://leetcode-cn.com/problems/partition-equal-subset-sum/solution/0-1-bei-bao-wen-ti-bian-ti-zhi-zi-ji-fen-ge-by-lab/
- [ ] 10. 算法题三，01背包变形，每个物品都有容量和价值，假设物品被分为两类，要求第一类必须买到k1个，第二类必须买k2个，每种物品只能买一次，给定容量的情况下求最大价值。
- [ ] 矩阵循环打印
----------------------------------------------------------
### **场景题**

- [ ] 1. 给出用户的数据和交易记录，是否给他开通花呗
- [ ] 2. 文本里面的“服务”、“食物好吃”等，如何抽取作为特征？（这问题当时理解有歧义，所以问了有点久才知道问的，词性关联模板）

>如何做好文本关键词提取？从三种算法说起 https://www.jiqizhixin.com/articles/2018-11-14-17

- [ ] 3. 给定一堆商品C，每个商品c对应有标题、描述，给定标签：时尚风、复古风、百搭风。设计技术方案，把每个商品c的主题标签给分类/聚类 等弄出来。没有训练数据，有哪些方案

- LDA主题建模,把所有标题和描述的文本放入LDA模型中进行主题聚类,
- Bert模型把标题和描述输入pre-train模型中得到句向量,用余弦相似度聚类句向量对应的商品
- 用tf-idf/n-grams提取词向量,用xgboost/rf进行分类
- 用word2vec模型生成的embedding把商品的文本转化为lstm的输入,训练出标题/描述-标签的模型 

- [ ]  4. 一个类似广告系统的应用问你怎么建模什么的
- [ ] 5. 对于团伙作案，从哪些维度考虑聚集性

> 反欺诈团伙识别 https://zhuanlan.zhihu.com/p/83395633
团伙欺诈发现一般涉及四类特征。1.基本信息特征，包括年龄，收入。2.环境信息，例如ip，设备。3.行为信息，消费或者操作信息。4.关联信息，个体之间的关联，例如转账什么的。

> 电商羊毛党套路多，半监督机器学习可以做什么？https://zhuanlan.zhihu.com/p/94709913
- 账号信息维度，如所有的账号注册日期相同、昵称相似（如均为*HL）、收获地址相似（如均为广东省深圳市龙岗区坂田街道A栋）等。
- 行为信息维度，如所有的交易金额总额相似（如均为200—209任意数字）、交易发生时间相近（如均为凌晨2:00—2:30任意时间）、交易路径相同（如均浏览并关注3家店铺后购买）等。
- 设备及网络信息维度，如所有设备的Wi-Fi信息相同（如Wi-Filist相同）、设备的分辨率相同、IP3相同、APP使用版本相同等。

- [ ] 对于无标签的数据，怎么考虑建模
> 电商羊毛党套路多，半监督机器学习可以做什么？https://zhuanlan.zhihu.com/p/94709913
- 半监督是介于有监督与无监督之间的机器学习，利用少量黑标签和无标签数据（通常黑标签与无标签数据万级，其中黑标签数据为千级），结合专家业务经验及数据挖掘技术，并通过特征工程全面提取欺诈特征，基于输入的特征组合进行聚类分析、图关联分析等，进而识别欺诈团伙。
- 半监督算法的核心在于欺诈特征的提取，即需要对账号信息、行为信息、设备及网络信息进行特征衍生，结合业务经验、数据挖掘技术等抽取高质量的特征并定义特征权重。
